### ISSUE TITLE
linac beam source

### ISSUE THREAD

**Issue by djboersma** [OpenGATE/Gate#72]
The linac beam source as implemented in `GateSourceLinacBeam.cc` needs work.

(1) It depends on a ROOT input file with a few thousand histograms. This ROOT input file can be generated from a phase space ROOT file, but the ROOT macro to do this conversion is not yet part of Gate. I wonder if we could/should maybe implement this as an actor instead of a separate macro.

(2) The histogram names (and lots of the comments in the code) are French, it would be good for international users and developers to change that to English.

(3) The code contains several hardcoded values that maybe should be configurable or dependent on input. For instance the number of volumes, currently set to 3.

(4) The code has several magic numbers for which it is unclear if they are specific to the linac for which the code was originally developed, or that they are more generic. E.g. this code snippet (in the current develop branch, lines 270-281) is particularly mysterious (I have no clue what the French comments are trying to tell me, and why there should be any linear relationship betwen the radius and the azimuthal angle):

```
  //DS TODO !!!

  if (volumeNumber==0) {Phi+=14.9*r/78.5;}             // G4cout<<"  Phi+=14.9*r/78.5= "<<Phi<< Gateendl;}
  if (volumeNumber==1) {Phi+=17.*r/80.;}               // G4cout<<"  Phi+=17.*r/80.= "<<Phi<< Gateendl;}
  if (volumeNumber==2) {Phi+=27.5*r/80.;}              // G4cout<<"  Phi+=27.5*r/80.= "<<Phi<< Gateendl;}
  //}
  //else {
  // on utilise une équation de droite représentative, qui est beaucoup plus juste (technique point source, pour chacun des 3 éléments)
  //if (volumeNumber==0) {Phi=14.9*r/78.5;}             // G4cout<<"  Phi+=14.9*r/78.5= "<<Phi<< Gateendl;}
  //if (volumeNumber==1) {Phi=17.*r/80.;}               // G4cout<<"  Phi+=17.*r/80.= "<<Phi<< Gateendl;}
  //if (volumeNumber==2) {Phi=27.5*r/80.;}              // G4cout<<"  Phi+=27.5*r/80.= "<<Phi<< Gateendl;}
  // }
```

We are using this actor in Uppsala and **maybe** I'll come around to fix these issues myself some time this summer.

**Comment by dsarrut** [comment]
to be discuss with Loic ;) !

**Comment by dsarrut** [comment]
I close it, considering the discussion #84

**Comment by djboersma** [comment]
I don't think that linac modelling can be done with GateSourceTPSPencilBeam, because an RT linac does not deliver a pencil beam and it does not do any scanning. So I think that these issues should be kept separate.

**Comment by dsarrut** [comment]
right, sorry I was too fast ;)

**Comment by GrevillotL** [comment]
Hello guys, I am back! I will try to help.
Priority one is French English Translation right?
David did you started?

**Comment by GrevillotL** [comment]
I meant Davis S obviously...

**Comment by djboersma** [comment]
David B here: French to English (item 2) is easiest to fix, I guess. But personally I think that it is more important to update the code to make it usable for other linacs than the one you modeled (item 3 and 4). If this actor is supposed to work for only one specific kind/brand/type of linac then it should not have such a generic name (`GateSourceLinacBeam`). If you do want to make it generic, then all components that can vary between linacs should be configurable and very well documented.

Also any external scripts/software should be provided or at least documented (item 1). You gave me the URL for the ROOT macros that you used during your PhD time, but we did not get around to change them in such a way that they would be useful to "any" linac user.

I worked a bit on all this when I was helping Hediye (guest researcher, she is back in Turkey for now) to do her linac modeling, but apart from that I have very little linac experience and I think that such experience is probably very important when redesigning the scripts and actors for a generic "linac beam". Also I am having my hands full with proton stuff. I am willing to brainstorm about this (in Strasbourg or in a skype call) to get an overview of how we should refactor this thing.

**Comment by GrevillotL** [comment]
Hi,
To put things back into context, I did all that as a Master sudent with a lot of support from a post-doc regarding the code - I was more on user level and I thought that my Supervisors would take care of the follow-up....
Currently like you, I am fully with ion therapy and therefore I cannot spend much efforts on that Topic, but I would like to help if I can - Translation may be a possibility not so time consuming... but allowing anyone to take over the current status of the source code and developments.

Regarding your questions, I think the source model is generic enough for all Linacs, but must be tested of course by new users like Hediye.

Questions (2) to (4) in your Initial message are related to the method developped in the published paper. I am not sure it would help changing the hard coded values or make everything user accessible. Again, it should be done if needed by new People taking over the Task and improving the modeling method, which I believe is generic.
